internal/buf/pool.go

```diff
@@ -0,0 +1,205 @@
+// Package buf manages allocation of temporary short-term buffers.
+package buf
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	maxAllocAttempts       = 3
+	allocationRetryTimeout = 500 * time.Millisecond
+)
+
+type segment struct {
+	nextUnallocated   int32  // high water mark
+	allocatedBufCount int32  // how many outstanding users of the segment there are
+	data              []byte // the underlying buffer from which we're allocating
+	pool              *Pool
+}
+
+// Buf represents allocated slice of memory pool. At the end of using the buffer, Release() must be called to
+// reclaim memory.
+type Buf struct {
+	nextUnallocated         int32
+	previousNextUnallocated int32
+
+	Data []byte
+
+	segment *segment // segment from which the data was allocated
+}
+
+// IsPooled determines whether data slice is part of a pool.
+func (b *Buf) IsPooled() bool { return b.segment != nil }
+
+// Release returns the slice back to the pool.
+func (b *Buf) Release() {
+	if b.segment == nil {
+		return
+	}
+
+	notify := false
+
+	// best effort compare-and-swap, which will pop the buffer off the stack in its appropriate segment
+	if atomic.CompareAndSwapInt32(&b.segment.nextUnallocated, b.nextUnallocated, b.previousNextUnallocated) {
+		// popped from the stack in the segment
+		notify = true
+	}
+
+	if atomic.AddInt32(&b.segment.allocatedBufCount, -1) == 0 {
+		// last allocated Buf, we can reset 'next' to zero
```

Comment:
I think there's a race here.
- Suppose the case where this is the last `Buf` in this segment.
- The last `Release` executes up to the line above, but not line 52.

In the mean time an allocate on the same same segment executes up to line 81, but not 82.
At this point `s.nextUnallocated` is after the `Buf.data` for this allocation

When 52 executes, it would reset `s.nextUnallocated` to 0 and it would corrupt the segment and subsequent allocations on the same segment.

Reply:
fixed by adding mutex and removing atomic